### DISCUSSION TITLE
How to set user-specific connection parameters.

### DISCUSSION THREAD

**Discussion by TOMATO-ONE** [neutrinolabs/xrdp#2410] [Ideas]
This is a discussion [derived](https://github.com/neutrinolabs/xrdp/discussions/2399#discussioncomment-4004633) from #2399. 

In the current xrdp.ini, connection parameters are common for all users.
Discussion continues on how to set user-specific connection parameters.

**Comment by TOMATO-ONE** [comment]
@metalefty has [an excellent idea f](https://github.com/neutrinolabs/xrdp/discussions/2399#discussioncomment-4054243)or a desire to change the IP address of the VNC/RDP Proxy connection on a per-user basis.
> It is an interesting use case for me, too. You don't need to withdraw the request so early but we can have a further discussion to find out better approaches. I think it can be achieved by a different approach. It's just an idea, I came up with hostname substitution.
> etalefty has an excellent idea for a desire to change the IP address of the VNC/RDP Proxy connection on a per-user basis.
> It is an interesting use case for me, too. You don't need to withdraw the request so early but we can have a further discussion to find out better approaches. I think it can be achieved by a differe
> In my idea, I would like to introduce some keywords for substitution. For example, if xrdp.ini is configured like this, `%u` will be replaced with the username.
> 
> ```
> [vnc-proxy]
> name=vnc-proxy
> lib=libvnc.so
> ip=%u.vnc.example.com.
> port=ask5900
> username=ask
> password=ask
> ```
> 
> Configure DNS like this to sort connections to the backend.
> 
> ```
> meta.vnc.example.com IN A 10.59.0.1
> matt.vnc.example.com IN A 10.59.0.2
> tomato-one.vnc.example.com IN A 10.59.0.2
> ```
> 
> This is how admins define different proxy destinations per user.

**Comment by matt335672** [comment]
I was thinking about this last night.

I think what we're doing here is simply adding a lookup capability to proxied neutrinordp and vnc connections. Is that right?

I'm sure the DNS idea will work, but there are some possible issues with it:-
1) What happens if the DNS name isn't matched? How do we report the error? Do you want to provide a fallback somehow?
2) Split admin responsibilities. In an enterprise environment it's possible the same admins may not be responsible for administering DNS and xrdp.

One way to address these might be to add a lookup capability to the `ip=` (and possibly `port=`) line(s). Something like:-

```
ip=LOOKUP("username", <filename>, <section>, <default>)
```

It's just a suggestion. Feel free to knock it down or add others.

**Reply by TOMATO-ONE** [reply]
Thank you.
I thought @metalefty's idea was an excellent idea to specify different connection hosts per user with only minor modifications to the current code.

@matt335672 , I acknowledge that there are some problems as you point out.
I thought I could alleviate that problem by putting the hostname in `/etc/hosts` on the xrdp host.

However, it is not a complete mechanism within xrdp.
I thought it would be better if a lookup table referencing mechanism such as you suggest could be achieved.

**Reply by TOMATO-ONE** [reply]
I thought @metalefty 's idea was a very good way to apply xrdp as a general purpose VNC/RDP Proxy.
Actually, it's not what I want to do, but I thought this method should go ahead because of its scalability.

I will raise this as a new issue once I have a direction for implementation.
